P5-164: Structure of code section (#111)
aau-hcl-p5/main: report/5.Solution/StructureCodebase.tex

\begin{itemize}
	\item Object Localization Algorithm
	\item Output / Communication
	\item Input / Webcam
\end{itemize}

These are all utilized by a controller class, called \texttt{FlatController}, which ties the solution together, and has the signature shown in Snippet~\ref{lst:FlatControllerInit}.
\begin{lstlisting}[language=Python,label={lst:FlatControllerInit},caption={Initialization method of the \texttt{FlatController} class}]
def __init__(self,
		algorithm: Callable[[np.ndarray], Vector],
		output_device: OutputDevice,
		input_device: VideoController,
		calibration_algorithm: Union[Callable[[], None], None] = None,
	) -> None:
	"""
	Initializes the controller
	
	:param algorithm: The algorithm to use for image processing
	:param output_device: The device to send data to
	:param input_device: What type the capturing device should be.
	:param calibration_algorithm: a function that takes calibration packages,
	and handles them in some unknown way (either logs them or sends them back)
"""

\end{lstlisting}

The order of the parameters does not correspond to the order of relevance, but rather is a requirement as two of the parameters have default arguments.

> This method takes, as the signature shows, a function, \texttt{algorithm} with an array as input and a vector as output.

Reviewer: den her sætning kunne jeg godt tænke mig omformuleret, mest fordi at jeg ikke er sikker på hvad den egentlig siger.

Måske kan den og de næste to slåes sammen til én sætning der bare fortæller hvad input og output er

This array is the array of pixels in a given frame.
The intent of this algorithm is to take a given frame, and localize the target within the frame.

\todo{implement the video output debug to actually be what i've written below}
The \texttt{output\_devices} parameter is a class that handles transmitting the location data to the NXT, but this can be overwritten to use a video feed as output as well, with the intent of debugging. 

> This argument can either be a singular \texttt{OutputDevice} or multiple, which is relevant when the user both wants to debug but also transmit the information to the NXT device.

Reviewer: ```suggestion
This argument can either be a single \texttt{OutputDevice} or multiple, which is relevant when the user both wants to debug while also transmitting the information to the NXT.
```

er stadig ikke sikker på hvad en output device er ud fra teksten

\todo{maybe change the capture\_type to also be either a class or a function, so we follow the same paradigm of coding all the way through}
Finally, the \texttt{capture\_type} parameter specifies which device the video feed should be received from.

> By default, this should be a camera feed, but it might be relevant to use a specific video file in some contexts, for instance when doing tests.

Reviewer: ```suggestion
By default, this is a camera feed, it can be changed to allow using a specific video file, which can be used when doing tests.
```

$ python3 main.py --help
usage: main.py [-h] [-a [name]] [-d]

Run the object detection part of the F.L.A.T system

optional arguments:
-h, --help            
	show this help message and exit
-a [name], --algorithm [name] 
	Choose which algorithm to run [GOTURN, YOLO, ZONE_AVG, OBJ_FILL]
	default='obj_fill'
-n [no_usb], --no-usb [no_usb]
	Whether to disable USB connection with the NXT.
	default=False
\end{lstlisting}
\todo{make sure these are still up to date when finishing the project, also image, bellow it isn't as Teknight created a new algorithm- but yeah}
Figure~\ref{fig:pythonClasses} shows the dependencies of each of the actual packages, and how each module is split up, into multiple packages or files.

\begin{figure}[H]
	\centering
	\includegraphics[width=\textwidth]{5.Solution/images/python_packages.png}
	\caption{The dependencies of the packages of the project{.} Image generated with pyreverse\cite{pyreverse}}
	\label{fig:pythonClasses}
\end{figure}

The project also contains functions for calibration, as seen on Figure~\ref{fig:pythonClasses}, which will be covered in Section~\ref{sec:calibration}.

\subsection{Movement}

> The codebase is not in the same way seperated into modules, in a direct way, due to the design of the language.

Reviewer: er det ikke super meget vrøvl?

> However the solution is divided into multiple file with distinct responsibilities.

Reviewer: ```suggestion
However, the solution is divided into multiple file with distinct responsibilities.
```

\begin{itemize}
	\item \texttt{tasks}
	\item \texttt{usb}
	\item \texttt{display}
	\item \texttt{movement}
	\item \texttt{calibration}
\end{itemize}

> the \texttt{tasks.c} can be seen as the primary file of the project, which creates all the tasks that the scheduling system enforces.

Reviewer: ```suggestion
The \texttt{tasks.c} can be seen as the primary file of the project, which creates all the tasks that the scheduling system enforces.
```

er den mere primær end de andre? den er vel bare et entry point ligesom main.py

> The code in snippet~\ref{lst:taskDecl} piece of code declares the tasks, but the specific aspects of each tasks, will be covered more indepth in section~\ref{sec:scheduling}

Reviewer: alt herfra og ned er outdated, virker fjollet at merge ind
